lib/multitimer/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+        // deadline after the timers have fired and only make one pass over the
+        // table.
+        let mut new_earliest = None;
+
+        for timer in self.timers.values_mut() {
+            // If the timer is on,
+            if let Some((d, r)) = timer.deadline {
+                // And the deadline has elapsed,
+                if d <= t {
+                    // Apply the repeat setting or disable the timer.
+                    if let Some(kind) = r {
+                        let next = match kind {
+                            Repeat::AfterWake(period) => t + period,
+                            Repeat::AfterDeadline(period) => d + period,
+                        };
+                        timer.deadline = Some((next, r));
+                    } else {
+                        timer.deadline = None;
+                    }
+                    // Record that it fired.
+                    timer.fired_but_not_observed = true;
+                }
+                // If the timer is _still_ on,
+                if let Some((new_d, _)) = timer.deadline {
+                    new_earliest = Some(if let Some(earliest) = new_earliest {
+                        new_d.min(earliest)
+                    } else {
+                        new_d
+                    });
+                }
+            }
+        }
+
+        sys_set_timer(new_earliest, 1 << self.notification_bit);
+    }
+
+    /// Checks all timer states unconditionally. This can be useful if you're
+    /// running a fast loop without waiting for notifications.
+    pub fn poll_now(&mut self) {
+        self.handle_notification(1 << self.notification_bit);
+    }
+
+    /// Returns an iterator over all timers that have fired since the last time
+    /// they were observed through this function. A timer may have fired more
+    /// than once; that information is lost.
+    ///
+    /// Timers that have fired will appear in the order given by their `Enum`
+    /// implementation, which in practice means declaration order.
+    ///
+    /// If you drop the iterator before it's exhausted, any timers you didn't
+    /// observe will appear next time you call this.
+    pub fn iter_fired(&mut self) -> impl Iterator<Item = E> + '_ {
+        self.timers.iter_mut().filter_map(move |(e, timer)| {
+            if core::mem::replace(&mut timer.fired_but_not_observed, false) {
+                Some(e)
+            } else {
+                None
+            }
+        })
+    }
+}
+
+trait SomeTimer {
+    fn get_timer(&self) -> u64;
+    fn set_timer(&self, deadline: Option<u64>, notifications: u32);
+}
+
+#[derive(Copy, Clone, Default)]
+pub struct Timer {
+    deadline: Option<(u64, Option<Repeat>)>,
+    fired_but_not_observed: bool,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Repeat {
+    AfterWake(u64),
+    AfterDeadline(u64),
+}
+
+// Syscall fakes for testing!
+
+#[cfg(not(target_os = "none"))]
```

Comment:
We could put all this into a single `mod fakes` (then import `super::fakes::*` in the test suite), to reduce repeated `#[cfg(not(target_os = "none"))]`

Reply:
Required a top-level conditional `use`, but, I think this is cleaner. Will change.